packages/context/src/context.ts

```diff
@@ -686,7 +734,36 @@ export class Context extends EventEmitter {
   findByTag<ValueType = BoundValue>(
     tagFilter: BindingTag | RegExp,
   ): Readonly<Binding<ValueType>>[] {
-    return this.find(filterByTag(tagFilter));
+    if (
+      tagFilter instanceof RegExp ||
+      (typeof tagFilter === 'string' &&
+        (tagFilter.includes('*') || tagFilter.includes('?')))
+    ) {
+      return this.find(filterByTag(tagFilter));
+    }
+    return this._findByTagIndex(tagFilter);
```

Comment:
While this solution works for the particular case of improving performance of controller methods when there are no interceptors configured, I am concerned it's too specialized and won't work for other scenarios. For example, if the application has 100 different interceptors bound in the context, to execute a controller method with no interceptors, we will still have to scan that array of 100 interceptors.

Also IIRC our previous conversations, we wanted to move away from `findByTag` in favor of more generic filtering. In this pull request, you are reversing that direction.

Have you considered implementing a more generic solution, one that will allow consumers outside of `@loopback/context` to implement their own cache and have an easy solution for invalidating it?

I think ideally, we want to:
- for each controller method, cache the actual list of interceptors to invoke 
- invalidate cache entries when the controller method metadata changes or when an interceptors is bound or unbound

Maybe we can treat this idea as a long-term goal and implement tag indexing as a short-term performance improvement.

Thoughts?

/cc @strongloop/loopback-maintainers

Reply:
I have added a few more commits to fully leverage `BindingTagFilter` is possible - even for `ContextView`.